util/monoclock.c

```diff
@@ -61,3 +61,36 @@ monoclock_get(struct timeval * tv)
 	/* Failure! */
 	return (-1);
 }
+
+/**
+ * monoclock_get_cputime(tv):
+ * Store how much time the process has been running in ${tv} if
+ * CLOCK_PROCESS_CPUTIME_ID is available; fall back to monoclock_get()
+ * otherwise.
+ */
+int
+monoclock_get_cputime(struct timeval * tv)
+{
+	struct timespec tp;
+
+	/* Use CLOCK_PROCESS_CPUTIME_ID if available. */
+#ifdef CLOCK_PROCESS_CPUTIME_ID
+	if (clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &tp) == 0) {
+		tv->tv_sec = tp.tv_sec;
+		tv->tv_usec = (suseconds_t)(tp.tv_nsec / 1000);
+	} else if ((errno != ENOSYS) && (errno != EINVAL)) {
```

Comment:
Umm, what happens if clock_gettime returns `ENOSYS`?

Reply:
?

Reply:
??

I just copied those lines from `monoclock_get()`.  `ENOSYS was removed in Issue 6:
http://pubs.opengroup.org/onlinepubs/9699919799/functions/clock_gettime.html
and there's no mention of it in the Linux or FreeBSD man pages.

Are you saying that if we get `ENOSYS` we should silently fall back to `monoclock_get()`?

Reply:
You failed at copying.  Right now if we get ENOSYS or EINVAL, this code silently returns 0.

But yes, we should fall back to monoclock_get if we get ENOSYS or EINVAL.  It's perfectly legal for the kernel to return EINVAL (we might e.g., have compiled using headers which are newer than the kernel) and ENOSYS is also sometimes returned (I can't remember exactly when, but I think it was Linux -- pretty sure there was an issue reported about that a couple years ago).

We should fall back to monoclock_get if CLOCK_PROCESS_CPUTIME_ID is not defined *or* if it fails with ENOSYS/EINVAL.